Feature/alec 178

BenjaminJ opened the pull request:
Sonar Cloud Security Grade A - Figure out What We Need to Fix and Report the List

BenjaminJ committed:
fix sonar security issue

BenjaminJ committed:
Sonarcloud - Make sure that expanding this archive file is safe here.

o Expanding archive files without controlling resource consumption is security-sensitive
https://sonarcloud.io/organizations/opennms/rules?open=java%3AS5042&rule_key=java%3AS5042

ALEC-178

BenjaminJ committed:
Sonarcloud security

- Make sure publicly writable directories are used safely here. java:S5443
- Generic exceptions should never be thrown

ALEC-178

BenjaminJ committed:
fix build

j-white commented on features/deeplearning/src/main/java/org/opennms/features/deeplearning/shell/LoadModel.java:
```diff
@@ -33,21 +33,25 @@
     public Object execute() {
         try (TFModel tfModel = new TFModel(modelPath)) {
-            System.out.println("Model successfully loaded: " + modelPath);
+            LOG.info("Model successfully loaded: {}",  modelPath);
```
We use `System.out` intentionally in Karaf shell commands so that they show up in console output. LOGs aren't ideal here.

j-white commented on engine/deeplearning/src/main/java/org/opennms/alec/engine/deeplearning/TFModel.java:
```diff
@@ -146,36 +148,36 @@ private void loadModelHyperParameters(String modelPath) {
             }
         } catch (IOException e) {
-            throw new RuntimeException("Failed to load model hyper parameters.", e);
+            throw new DeepLearningException("Failed to load model hyper parameters.", e);
```
I'd opt to avoid adding a new checked exception here. This add a bunch of boilerplate, and can now lead to an NPE in calls to createEngine.

j-white commented on engine/dbscan/src/main/java/org/opennms/alec/engine/dbscan/DBScanEngineFactory.java:
```diff
@@ -72,7 +72,7 @@ public AbstractClusterEngine createEngine(MetricRegistry metrics) {
             return new DBScanEngine(metrics, epsilon, alpha, beta, factory.get());
         } else {
-            LOG.error("Wrong distance measure configuration {}, we'll use default {}", distanceMeasureFactoryName, alarmInSpaceAndTimeDistanceMeasureFactory.getName());
+            LOG.error("Warn distance measure configuration {}, we'll use default {}", distanceMeasureFactoryName, alarmInSpaceAndTimeDistanceMeasureFactory.getName());
```
`Wrong` is more appropriate than `Warn` here.

BenjaminJ replied:
warn was intended for the log level, thank you to have spotted that

j-white commented on driver/test/src/main/java/org/opennms/alec/driver/test/TestDriver.java:
```diff
@@ -265,13 +286,13 @@ private void printTick(long tick, long firstTimestamp, long lastTimeStamp, long
             return;
         }
         double percentageComplete = ((tick - firstTimestamp) / (double)(lastTimeStamp - firstTimestamp)) * 100d;
-        System.out.printf("Tick at %s (%d) - %.2f%% complete - %s elapsed\n", new Date(tick), tick,
```
The usage of `System.out` is intentional in this class, since it's used in Karaf shell commands.

BenjaminJ committed:
Sonarcloud security

- use of System.out intentionally in Karaf shell commands so that they show up in console output. LOGs aren't ideal here.
- avoid adding a new checked exception here. This add a bunch of boilerplate, and can now lead to an NPE in calls to createEngine.
- use java.io.tmpdir to avoid sonar java:S5443 (Make sure publicly writable directories are used safely here.)

ALEC-178

BenjaminJ committed:
revert change

BenjaminJ committed:
Sonarcloud security

ALEC-178

amay94 commented on engine/deeplearning/src/main/java/org/opennms/alec/engine/deeplearning/ClasspathUtils.java:
```diff
@@ -97,19 +105,56 @@ static void copyResourcesRecursively(URL originUrl, File destination) throws IOE
+
+                while ((nBytes = in.read(buffer)) > 0) {
+                    if (ze.getName().startsWith(jarConnection.getEntryName())) {
+                        final String fileName = removeStart(ze.getName(), jarConnection.getEntryName());
```
It looks like this could be done before the while loop. Also, if all of the zip entries are retrieved from the jarConnection, is this if statement needed?

BenjaminJ replied:
you're right, fixed

amay94 commented on engine/deeplearning/src/main/java/org/opennms/alec/engine/deeplearning/ClasspathUtils.java:
I think this comment is on the wrong variable

BenjaminJ replied:
also fixed

amay94 reviewed (dismissed):
Looks good to me other than a few minor comments

BenjaminJ committed:
Sonarcloud security

o fix comment at wrong line
o filename could be set before the loop
o all the zip entries are retrieved from the jarConnection, the if statement is redundant

ALEC-178

amay94 reviewed (approved):
LGTM. Thanks for the updates
